### PR TITLE
feat(analytics): emit swap failure events and reject same-token pairs

### PR DESCRIPTION
**Pull request by alastairong1**
## Why

A swap client that fails **100% of the time** is currently indistinguishable from one sending no traffic at all.

- `swap_quoted` / `swap_calldata_generated` fire only on success
- `api_request` fires on failures but carries no request body

So a broken integration produces zero diagnosable signal.

This surfaced in production: an integrator (`terminal1`) spent **five days getting a 404 on every single `/v1/swap/quote` call** — 177 attempts, 0 successes — and nobody noticed, because the Growth "Demand" tab reads `swap_quoted` and therefore saw them as *absent* rather than *failing*.

Working out the cause required matching latency fingerprints against manually replayed requests, since nothing recorded what they sent:

| Client | 404s | p50 latency |
|---|---:|---:|
| terminal1 | 177 | 14,009 ms |
| st0x.io | 43 | 1,079 ms |

161 of terminal1's 177 failures sit in a 12–20s band. Replaying all 34 tokens in both directions found exactly one request shape that reproduces it — `USDC → USDC`, at 14.9–18.2s. Everything else either succeeds or fails in under 1.1s.

## What

**1. Failure events on all four swap endpoints.** `swap_quote_failed` and `swap_calldata_failed` carry the requested pair, amount, denomination, mode, `api_version`, and the wire `error_code` / `status_code`.

- `same_token` is precomputed, because that failure mode is invisible in aggregate — spotting it otherwise means grouping on two address columns, which is exactly the step nobody takes.
- Success and failure events stay disjoint, so a per-client success rate is well defined and alertable.
- Calldata attribution failures feed the same event: the caller ends up with no calldata either way.

**2. Reject `inputToken == outputToken` up front,** with a new `SWAP_SAME_TOKEN` code (HTTP 400).

Such a request is always a caller bug, but not a cheap one. The token is individually supported so `validate_supported_tokens` passes, and the pair matches every order holding that token — for a stablecoin leg that is the entire book, which then gets quoted over RPC before the simulation finds no executable leg and returns "no liquidity". A ~15s round trip answering a malformed question with a misleading answer. It is also a cheap amplification vector: one request costs a full-book RPC sweep.

Measured, via the site proxy:

| Request | Result |
|---|---|
| `USDC → USDC` (20 orders) | 404 in ~15,000 ms |
| `wtTSLA → wtTSLA` (3 orders) | 404 in ~1,072 ms |
| `USDC → wt`, 34 tokens | 32 × 200 in ~1.0–1.6s |
| `wt → USDC`, 34 tokens | 33 × 200, max 3.8s |

`RaindexError::SameTokenPair` also now maps to `SWAP_SAME_TOKEN` instead of being flattened into the generic invalid-parameters bucket, so it stays diagnosable if it ever surfaces from deeper in the SDK.

**3. `ApiError::code()` / `public_message()` extracted** so the HTTP responder and analytics report an error identically — a drift there would let a dashboard disagree with what the caller actually received.

## Notes for review

- **No new PII.** Token addresses are public contract addresses and amounts are trade sizes — the same class of data the success events already emit. `taker` appears only on calldata failures, and only where the success event already carried it.
- **Negligible volume.** Failures are rare (177 + 43 over five days).
- The `SwapFailure` snapshot is gated on `analytics.is_enabled()`, so a disabled sink still builds nothing.

## Testing

375 tests pass (12 new). `nix develop -c rainix-rs-static` (clippy + fmt) is clean.

New coverage: failure events fire on the error path with the right pair/code/status for v1 and v2; success and failure remain disjoint; the guard rejects a same-token pair *before* touching the data source (proved with a data source that errors on first use); `SWAP_SAME_TOKEN` serialization and 400 mapping; `code()`/`public_message()` across every `ApiError` variant.

## Follow-ups (not in this PR)

- A PostHog alert on per-client quote success rate — the point of the new events.
- Terminal1 still needs telling; this PR makes the next occurrence visible, it does not fix their integration.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
<a href="https://app.blacksmith.sh/ST0x-Technology/codesmith/st0x.rest.api/pr/167"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1787907382&installation_model_id=19370&pr_number=167&repository=ST0x-Technology%2Fst0x.rest.api&return_to=https%3A%2F%2Fgithub.com%2FST0x-Technology%2Fst0x.rest.api%2Fpull%2F167&signature=e1967793dedf8224b60eb3e466608be4170302ae78e94efeafb4216c89ae8a0e"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Same-token swap requests are rejected early with a clear `SWAP_SAME_TOKEN` error and HTTP 400 response.
  * Validation applies consistently to quote and calldata requests across supported API versions.
  * Error responses now provide more consistent public error codes and messages.

* **Analytics**
  * Failed quote and calldata requests now include relevant request, API-version, and error details.
  * Successful swap events now include API-version metadata.

* **Documentation**
  * Updated analytics and error documentation with supported events, privacy handling, and error codes.

### CODE REVIEW

**Comment by findolor** [comment]
* **#167** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.rest.api/167?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.rest.api/167?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>add-to-gt-merge-queue</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/ST0x-Technology?ref=merge-queue-instructions-in-stack-comment&prId=167).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/routes/swap/calldata.rs [hunk 1]
```diff
@@ -156,13 +158,49 @@ async fn handle_swap_calldata(
             req.taker,
             req.input_token,
             req.output_token,
+            req.output_amount.clone(),
             serde_json::to_value(req.denomination).unwrap_or(serde_json::Value::Null),
         )
     });
```
Comment (JuaniRios) [inline src/routes/swap/calldata.rs]: should fix: `analytics_context` is an unnamed tuple, and three of its fields have the `Address` type. The code destructures it by position at three separate sites. If one site swaps `input_token` and `output_token`, the code still compiles, and the failure events report the wrong pair. Replace the tuple with a small struct that has named fields. Do the same for the 6-tuple in `handle_swap_calldata_v2`.

File: src/routes/swap/quote.rs [hunk 2]
```diff
@@ -246,6 +320,8 @@ async fn process_swap_quote_v2(
     ds: &dyn SwapDataSource,
     req: SwapQuoteV2Request,
 ) -> Result<SwapQuoteV2Response, ApiError> {
+    ensure_distinct_tokens(req.input_token, req.output_token)?;
```
Comment (JuaniRios) [inline src/routes/swap/quote.rs:282]: should fix: this guard closes one expensive 400, but a cheaper one stays open in the same function. The `price_cap` / `slippage_bps` combination is validated only after `get_orders_for_pair` and `build_candidates_for_pair`. A request with an invalid combination runs the full order query and the RPC sweep, then gets a 400. When the request has a non-zero `taker`, the candidates cache is bypassed, so each such request is a full uncached sweep. Move the match arms that only return `BadRequest` up here, before the data-source calls. The v2 calldata path already validates in `TryFrom` before it touches the data source.

File: src/analytics/events.rs [hunk 3]
```diff
@@ -200,6 +204,105 @@ pub(crate) fn swap_calldata_generated_event(
...
+        props.insert(
+            "requested_amount".to_string(),
+            failure.requested_amount.into(),
+        );
+    }
+    props.insert("denomination".to_string(), failure.denomination);
+    props.insert(
+        "api_version".to_string(),
+        failure.api_version.as_str().into(),
+    );
```
Comment (JuaniRios) [inline src/analytics/events.rs:277]: minor: the failure events carry `api_version`, but `swap_quoted_event` and `swap_quoted_v2_event` both emit `swap_quoted` without it. A per-version success rate needs the same key on both sides. Add `api_version` to the two quote success events, the same as `swap_calldata_generated_event` does.

File: src/routes/swap/quote.rs [hunk 4]
```diff
@@ -890,6 +967,146 @@ mod tests {
...
+        let result = handle_swap_quote(
+            &ds,
+            &test_key(),
+            &Analytics::disabled(),
+            quote_request("100"),
+        )
+        .await;
+
+        assert_error_code(result, ApiErrorCode::SwapNoLiquidity);
+    }
```
Comment (JuaniRios) [inline src/routes/swap/quote.rs]: minor: this test cannot fail for the reason the doc comment gives. `Analytics::disabled()` exposes nothing observable, and the only assertion is the error code, which an earlier test already checks. If someone removes the `is_enabled` gate, this test still passes. Either extract the snapshot into a function and assert that it returns `None` for disabled analytics, or delete the test.

File: src/routes/swap/quote.rs [hunk 5]
```diff
@@ -127,7 +130,42 @@ async fn handle_swap_quote(
...
+    // Snapshot the request before it is consumed: on the failure path there is no
+    // response to describe, and the request is the only record of what was asked for.
+    let failure_context = analytics.is_enabled().then(|| {
+        (
+            req.input_token,
+            req.output_token,
+            req.output_amount.clone(),
+            serde_json::to_value(req.denomination).unwrap_or(serde_json::Value::Null),
+        )
+    });
```
Comment (JuaniRios) [inline src/routes/swap/quote.rs]: minor: this snapshot only runs after Rocket's JSON guard accepts the body. A client that sends a malformed body gets a 422, and neither `swap_quoted` nor `swap_quote_failed` fires. That client still shows as absent, which is the failure mode this PR describes. Either document that the success-rate alert must be paired with an `api_request` alert on `status_code >= 400` for these endpoints, or accept `Json<serde_json::Value>` and deserialize inside the handler, so the 422 path also emits a failure event.

File: src/routes/swap/mod.rs [hunk 6]
```diff
@@ -355,6 +393,32 @@ mod tests {
         serde_json::from_value(value).expect("deserialize mock order")
     }
 
+    #[test]
+    fn test_ensure_distinct_tokens_allows_a_real_pair() {
+        let usdc = address!("833589fCD6eDb6E08f4c7C32D4f71b54bdA02913");
+        let weth = address!("4200000000000000000000000000000000000006");
+        assert!(ensure_distinct_tokens(usdc, weth).is_ok());
```
Comment (JuaniRios) [inline src/routes/swap/mod.rs:459]: nit: this positive case is the trivial complement of a two-line equality check, and the reject test next to it already pins the interesting branch. Keep it as cheap insurance, or fold both cases into one test.

File: src/analytics/events.rs [hunk 7]
```diff
@@ -200,6 +204,105 @@ pub(crate) fn swap_calldata_generated_event(
...
+
+/// Keep arbitrary request text out of PostHog while preserving valid trade sizes.
+///
+/// Swap amount fields deserialize as strings and failure events include parse errors,
+/// so a failed request is not proof that the value is numeric. The length bound also
+/// keeps an authenticated caller from creating oversized analytics properties.
+fn analytics_requested_amount(amount: &str) -> bool {
+    amount.len() <= MAX_ANALYTICS_AMOUNT_LENGTH && Float::parse(amount.to_string()).is_ok()
+}
+
```
Comment (JuaniRios) [inline src/analytics/events.rs:308]: minor: no test exercises the length branch of this check. The new tests only cover the numeric-parse branch with a short string. Add a test with a valid numeric string longer than 128 bytes, and a boundary case at exactly 128 bytes.

File: src/error.rs [hunk 8]
```diff
@@ -291,11 +316,65 @@ mod tests {
         assert!(logs_contain("SWAP_NO_LIQUIDITY"));
     }
 
+    /// `code()`/`public_message()` back both the HTTP response and the analytics
+    /// event, so a drift here would let PostHog disagree with what the caller saw.
+    #[test]
+    fn test_code_and_message_cover_every_variant() {
```
Comment (JuaniRios) [inline src/error.rs]: nit: seven of the eight cases assert that a 1:1 named mapping returns itself. Only the `Coded` case can plausibly be miswired. Trim the table to the `Coded` case plus one representative variant, or replace it with a check that one request yields the same `error_code` in the response body and in the captured analytics event.

**Review by JuaniRios — CHANGES_REQUESTED** [review]

**Review by JuaniRios — APPROVED** [review]
